rootston/output.c

```diff
@@ -53,46 +53,8 @@ static void render_surface(struct wlr_surface *surface,
 	};
 	if (wlr_output_layout_intersects(desktop->layout, wlr_output, &render_box)) {
 		float matrix[16];
-
-		float translate_center[16];
-		wlr_matrix_translate(&translate_center,
-			(int)ox + render_width / 2, (int)oy + render_height / 2, 0);
-
-		float rotate[16];
-		wlr_matrix_rotate(&rotate, rotation);
-
-		float translate_origin[16];
-		wlr_matrix_translate(&translate_origin, -render_width / 2,
-			-render_height / 2, 0);
-
-		float scale[16];
-		wlr_matrix_scale(&scale, render_width, render_height, 1);
-
-		float transform[16];
-		wlr_matrix_mul(&translate_center, &rotate, &transform);
-		wlr_matrix_mul(&transform, &translate_origin, &transform);
-		wlr_matrix_mul(&transform, &scale, &transform);
-
-		if (surface->current->transform != WL_OUTPUT_TRANSFORM_NORMAL) {
-			float surface_translate_center[16];
-			wlr_matrix_translate(&surface_translate_center, 0.5, 0.5, 0);
-
-			float surface_transform[16];
-			wlr_matrix_transform(surface_transform,
-				wlr_output_transform_invert(surface->current->transform));
-
-			float surface_translate_origin[16];
-			wlr_matrix_translate(&surface_translate_origin, -0.5, -0.5, 0);
-
-			wlr_matrix_mul(&transform, &surface_translate_center,
-				&transform);
-			wlr_matrix_mul(&transform, &surface_transform, &transform);
-			wlr_matrix_mul(&transform, &surface_translate_origin,
-				&transform);
-		}
-
-		wlr_matrix_mul(&wlr_output->transform_matrix, &transform, &matrix);
-
+		wlr_output_get_box_matrix(wlr_output, ox, oy, render_width,
+			render_height, surface->current->transform, rotation, &matrix);
```

Comment:
:+1:

Reply:
@emersion do you like this abstraction?

Reply:
* Make it take a `wlr_box` instead of four parameters
* Coordinates are in output-local coordinates (multiplied by `output.scale`), if it's an output function I'd expect coordinates to be in surface-local coordinates and the function to take care of multiplying (as it already takes care of `output->transform`)
* This function has nothing to do with the output, apart from multiplying the result to the output matrix at the end. I'd prefer this to be a `wlr_matrix` function.
* Rotation is opiniated and rootston-specific

Reply:
>Make it take a wlr_box instead of four parameters

:+1: 

>Coordinates are in output-local coordinates (multiplied by output.scale), if it's an output function I'd expect coordinates to be in surface-local coordinates and the function to take care of multiplying (as it already takes care of output->transform)

:-1: 

>Rotation is opiniated and rootston-specific

Not really. This function appears to me to be box-oriented rather than surface-oriented and rotating boxes isn't particularly opinionated.

Reply:
>This function has nothing to do with the output, apart from multiplying the result to the output matrix at the end. I'd prefer this to be a wlr_matrix function.

I'd need to see the proposed function signature of it as it appears in wlr_matrix.

Reply:
>I'd need to see the proposed function signature of it as it appears in wlr_matrix.

```c
void wlr_matrix_box_transform(float (*mat)[16], struct wlr_box *box,
	enum wl_output_transform transform, float rotation);
```

Still searching for a better name.

Reply:
Maybe it'd be better on wlr_box rather than wlr_matrix? `wlr_box_transform`

Reply:
This name seems like it transforms a box to another box. I'd like to have `wlr_matrix`, `box` and `transform` in the name. Maybe `wlr_matrix_transform_box`? `wlr_matrix_from_transformed_box`?

Reply:
`wlr_box_projection`?

Reply:
That sounds better. But this function really belongs to `render/matrix.c` IMHO. `wlr_matrix_project_box`? `wlr_matrix_projection`?

Reply:
Doesn't this matrix have an official name? Is this a "model" matrix or a "view" matrix?

Reply:
I'm not sure that the lingo maps that precisely to our circumstances.

Reply:
I don't think we can use wlr_box for this as-is because it has `x` and `y` in integers but we need it to be a double or a float to match the output coordinates.

Reply:
That's not true. Coordinates are output-locale, which means using a double will allow rendering a view between two pixels. We had blurry rendering because of this in the past.